auth_fastapi/session.py

```diff
@@ -0,0 +1,47 @@
+import edgedb
+
+from typing import Annotated, Optional, Union
+from fastapi import Cookie, Depends
+
+ClientDep = Annotated[edgedb.AsyncIOClient, Depends(edgedb.create_async_client)]
+
+
+class BaseSession:
+    client: edgedb.AsyncIOClient
+
+    def __init__(self, *, client: edgedb.AsyncIOClient):
+        self.client = client
+
+    async def is_authenticated(self) -> bool:
+        return await self.client.query_required_single(  # type: ignore
+            "select exists ext::auth::ClientTokenIdentity"
+        )
+
+class AuthenticatedSession(BaseSession):
+    auth_token: str
+
+    def __init__(self, *, client: edgedb.AsyncIOClient, auth_token: str):
+        self.auth_token = auth_token
+        self.client = client.with_globals(  # type: ignore
+            {"ext::auth::ClientTokenIdentity": auth_token}
+        )
+
+
+class AnonymousSession(BaseSession):
+    pass
+
+
+Session = Union[AuthenticatedSession, AnonymousSession]
+
+
+def extract_session(
+    auth_token: Annotated[Optional[str], Cookie(alias="edgedb_auth_token")],
+    client: ClientDep,
```

Comment:
I think I don't understand the dependency injection system well enough to pull this off, but my intention here is to declare a dependency on a service that provides a `edgedb.AsyncIOClient`, but _not_ need to define it explicitly via `create_async_client` and have the downstream consumer of this library fill it in with their own function that returns a client instance that might have special configuration.

Reply:
Right, that's what I usually do in production. Making the `FastAPI` application a constructor also helps with testing. I'll try to modify this sample and see where it hits the wall.